### PR TITLE
add dnspolicy

### PR DESCRIPTION
**Pull request by maleck13**
@alexsnaps this PR moves the content over from the MGC repo to this one for DNSPolicy. Ok to merge?

### CODE REVIEW

**Review by alexsnaps — COMMENTED** [review]
Most of it (under `## Proposal` feels like the `# Guide-level explanation` afaict?
Also, there @guicassolato had some comments on using a "more standard" label selector.

Anyways, on the merging, the idea is that RFC is full fledged proposal when merging. So since this is all still under "proposal", do we think that is (to the best of our current knowledge) the feature as we'd expose it to users? If so that'd cover the `Guide-level explanation` section... `Reference-level explanation` part of _how_ it'd be implemented isn't really much spelled out, is it?

**Comment by maleck13** [comment]
Yes this has been updated as we have implemented the API so it should be as you say upto date for the guide level. How important  is the reference level?
Often things get implemented in one way and then things change and get refactored etc so I worry this could get out of date rapidly. I could keep it at a high level IE

- Add DNSPolicy CRD that conforms to policy attachment spec
- Add DNSPolicy controller to MCG
- DNS logic and record management should all migrate out of the gateway controller into this new controller as it is the responsibility and domain of the DNSPolicy controller to manage DNS

File: rfcs/0000-dns-policy.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
+# RFC Template
+
+- Feature Name: DNSPolicy
+- Start Date: 2023-07-01
+- RFC PR: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/pull/0000)
+- Issue tracking: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/issues/0000)
```
[the remark this answers is not in the document]
Comment (maleck13) [reply rfcs/0000-dns-policy.md]: updated with the best issue for it

File: rfcs/0000-dns-policy.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
+# RFC Template
+
+- Feature Name: DNSPolicy
+- Start Date: 2023-07-01
+- RFC PR: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/pull/0000)
```
[the remark this answers is not in the document]
Comment (maleck13) [reply rfcs/0000-dns-policy.md]: updated

**Comment by maleck13** [comment]
Moved main body of text to guide-level
added a high level reference

**Comment by alexsnaps** [comment]
@guicassolato wdyt? Merge as is?

**Comment by maleck13** [comment]
@guicassolato @alexsnaps as this a migration from the MGC repo wondering if we can merge at this point or is there something fundamental that needs to be addressed. IE we have already implemented it and documented it so I think any change would be a new RFC possibly?

**Review by alexsnaps — APPROVED** [review]
Rename this to `rfcs/0003-dns-policy.md` and let's merge this.

**Comment by maleck13** [comment]
done

**Comment by maleck13** [comment]
@alexsnaps @guicassolato woops :) fixed now

**Comment by maleck13** [comment]
@alexsnaps I have done all that was needed I think now, ok to merge?

**Review by alexsnaps — APPROVED** [review]

**Comment by maleck13** [comment]
@guicassolato I made the requested change

File: rfcs/0003-dns-policy.md [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+The DNSPolicy also covers health checks that inform the DNS response but that is not covered in this document.
+
+Below is a draft API for what we anticipate the DNSPolicy to look like
+
+```yaml 
+
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSPolicy
+spec:
+  targetRef: # defaults to gateway gvk and current namespace
```
Comment (guicassolato) [inline rfcs/0003-dns-policy.md:43]: GVK is required, I'm afraid. Only namespace can be omitted (and assumed.) https://github.com/kubernetes-sigs/gateway-api/blob/7042cd866bc69b160ab6d04b51676e14b487505d/apis/v1alpha2/policy_types.go#L24-L41

**Review by guicassolato — APPROVED** [review]
Good base ref for the work already started around DNSPolicy.

Small comment left about a minor detail required to comply with GW-API PA types that, as of today, would not allow implementing exactly as exemplified, but it's certainly not the main focus of the RFC here, neither its impact on UX is of concern IMO.

I missed something about hostnames specified only at the level of the HTTPRoutes (subdomains), for which one may still want to set LB rules. My understanding is that, while limited to gateways, those would have to be brought up as additional listener definitions, whose limitations are known. I suppose this could be added in the [Future Possibilities](https://github.com/Kuadrant/architecture/blob/eddfb3cf43299d90dd13726a0a8b74c569a3009a/0000-template.md?plain=1#L74) section or in a future iteration.

**Comment by maleck13** [comment]
Thanks @guicassolato yes I think section name support will be something we will look at in the  near future

**Pull request merged**